/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2023-2026 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : Add Vendor specific commands.
*                       : Add Vendor specific commands for the Independent Reset & the set
*                         EUI-64 MAC address features.
*                       : Add Vendor specific commands for set and get  TX power limit feature.
*                       : Add Vendor specific commands for RF test mode (MFG commands).
*                       : Add Vendor specific commands for configuring CCA mode and threshold at run time.
*                       : Add Vendor specific commands for Reading the FW version.
*                       : Add Vendor specific commands for upgrading the independant reset feature for OOB
*                         GPIO soft reset.
* Version               : v007
*                       : v004 -Modified TX power limit VSC to support enable/disable TX Power clamping on CH 26
*                       : v005 -Return Invalid argument error if TX power limit > 127.
*                       : v006 -Ported spifreq VSC command, SPI mfg command and generic MFG support from 
*                         individual patch files.
*                       : v007 -Openthread rebase
* Mandatory/Optional    : Optional
* Impact if not applied : VSC commands won't be available.
*/

Upstream-Status: Inappropriate [platform specific changes]

diff --git a/include/openthread/platform/radio.h b/include/openthread/platform/radio.h
index c93458e95..46aa02bba 100644
--- a/include/openthread/platform/radio.h
+++ b/include/openthread/platform/radio.h
@@ -2,6 +2,8 @@
  *  Copyright (c) 2016, The OpenThread Authors.
  *  All rights reserved.
  *
+ *  Copyright 2021-2024 NXP.
+ *
  *  Redistribution and use in source and binary forms, with or without
  *  modification, are permitted provided that the following conditions are met:
  *  1. Redistributions of source code must retain the above copyright
@@ -475,6 +477,25 @@ typedef struct otLinkMetrics
     bool mReserved : 1;   ///< Reserved, this is for reference device.
 } otLinkMetrics;
 
+/**
+ * This structure represents different CCA mode configurations before Tx.
+ */
+typedef struct otCCAModeConfig
+{
+    uint8_t mCcaMode;          ///< CCA Mode type [CCA1=0x01, CCA2=0x02, CCA3=0x03[CCA1 AND CCA2], CCA3=0x04[CCA1 OR CCA2], NoCCA=0xFF.
+    uint8_t mCca1Threshold;    ///< Energy threshold for CCA Mode1.
+    uint8_t mCca2CorrThreshold;    ///< CCA Mode 2 Correlation Threshold.
+    uint8_t mCca2MinNumOfCorrTh;    ///< CCA Mode 2 Threshold Number of Correlation Peaks.
+} otCCAModeConfig;
+
+/**
+* This structure represents threshold time for OOB 15.4 soft reset.
+*/
+typedef struct otIRConfig
+{
+    uint16_t mIRThreshold;    ///< threshold time in ms at which OOB toggle will trigger an IR
+} otIRConfig;
+
 /**
  * @}
  */
@@ -777,6 +798,18 @@ uint64_t otPlatRadioGetNow(otInstance *aInstance);
  */
 uint32_t otPlatRadioGetBusSpeed(otInstance *aInstance);
 
+/**
+ * Set the bus speed in bits/second between the host and the radio chip.
+ *
+ * @param[in]   aInstance    A pointer to an OpenThread instance.
+ * @param[in]   aSpeed       The SPI Bus Speed request.
+ *
+ * @returns The bus speed in bits/second between the host and the radio chip.
+ *          Return 0 when the MAC and above layer and Radio layer resides on the same chip.
+ *
+ */
+otError otPlatRadioSetBusSpeed(otInstance *aInstance, uint32_t aSpeed);
+
 /**
  * Get the bus latency in microseconds between the host and the radio chip.
  *
@@ -1437,6 +1470,65 @@ extern otError otPlatRadioGetRawPowerSetting(otInstance *aInstance,
                                              uint8_t     aChannel,
                                              uint8_t    *aRawPowerSetting,
                                              uint16_t   *aRawPowerSettingLength);
+/**
+ * @}
+ *
+ */
+otError otPlatRadioSetIRCmd(otInstance *aInstance);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioSetIRConfig(otInstance *aInstance, uint8_t mode);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioGetIRConfig(otInstance *aInstance, uint8_t* mode);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioSetIeeeEui64(otInstance *aInstance, const otExtAddress *aIeeeEui64);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioSetTxPowerLimit(otInstance *aInstance, uint8_t txPowerLimit);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioGetTxPowerLimit(otInstance *aInstance, uint8_t* txPowerLimit);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioMfgCommand(otInstance *aInstance, uint8_t *payload, const uint8_t payloadLenIn, uint8_t *payloadLenOut);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioCcaConfigValue(otInstance *aInstance, otCCAModeConfig *aCcaConfig, uint8_t aSetValue);
+
+/**
+ * @}
+ *
+ */
+otError otPlatRadioGetFwVersionString(otInstance *aInstance, const char *fwVersion, uint8_t fwVersionLen);
+
+/**
+ * @}
+ *
+*/
+otError otPlatRadioIRThresholdConfig(otInstance *aInstance, otIRConfig *aIRConfig, uint8_t aSetValue);
 
 /**
  * @}
diff --git a/src/lib/spinel/radio_spinel.cpp b/src/lib/spinel/radio_spinel.cpp
index 5d394b10e..a070545e7 100644
--- a/src/lib/spinel/radio_spinel.cpp
+++ b/src/lib/spinel/radio_spinel.cpp
@@ -937,6 +937,111 @@ otError RadioSpinel::GetIeeeEui64(uint8_t *aIeeeEui64)
     return OT_ERROR_NONE;
 }
 
+otError RadioSpinel::SetIeeeEui64(const otExtAddress &aIeeeEui64)
+{
+    otExtAddress addr;
+    otError      error;
+
+    for (size_t i = 0; i < sizeof(addr); i++)
+    {
+        addr.m8[i] = aIeeeEui64.m8[sizeof(addr) - 1 - i];
+    }
+
+    SuccessOrExit(error = Set(SPINEL_PROP_VENDOR_NXP_SET_EUI64_CMD, SPINEL_DATATYPE_EUI64_S, addr.m8));
+    sIeeeEui64 = aIeeeEui64;
+exit:
+    return error;
+}
+
+otError RadioSpinel::SetTxPowerLimit(uint8_t txPowerLimit)
+{
+    otError error;
+    SuccessOrExit(error = Set(SPINEL_PROP_VENDOR_NXP_GET_SET_TXPOWERLIMIT_CMD, SPINEL_DATATYPE_UINT8_S, txPowerLimit));
+
+exit:
+    return error;
+}
+
+otError RadioSpinel::GetTxPowerLimit(uint8_t &txPowerLimit)
+{
+    otError error = Get(SPINEL_PROP_VENDOR_NXP_GET_SET_TXPOWERLIMIT_CMD, SPINEL_DATATYPE_UINT8_S, &txPowerLimit);
+    return error;
+}
+
+otError RadioSpinel::SetIRConfig(uint8_t mode)
+{
+    otError error = Set(SPINEL_PROP_VENDOR_NXP_IR_CONFIG, SPINEL_DATATYPE_UINT8_S, mode);
+    return error;
+}
+
+otError RadioSpinel::GetIRConfig(uint8_t &mode)
+{
+    otError error = Get(SPINEL_PROP_VENDOR_NXP_IR_CONFIG, SPINEL_DATATYPE_UINT8_S, &mode);
+    return error;
+}
+
+#define MAXBUFFERSIZE 16
+
+otError RadioSpinel::MfgCmd(uint8_t *payload, const uint8_t payloadLenIn, uint8_t &payloadLenOut)
+{
+    otError error;
+    uint8_t buffer[MAXBUFFERSIZE]; //temporary buffer used to be passed as arg of GetWithParam
+    spinel_ssize_t packed;
+
+    packed = spinel_datatype_pack(buffer, sizeof(buffer), SPINEL_DATATYPE_DATA_S, payload, payloadLenIn);
+    error = GetWithParam(SPINEL_CMD_VENDOR_NXP_MFG, buffer, static_cast<spinel_size_t>(packed), SPINEL_DATATYPE_DATA_S, payload, &payloadLenOut);
+
+    return error;
+}
+
+otError RadioSpinel::CcaConfigValue(otCCAModeConfig &aCcaConfig, uint8_t aSetValue)
+{
+    otError error;
+    uint8_t aCcaMode, aCca1Threshold, aCca2CorrThreshold, aCca2MinNumOfCorrTh;
+
+    if(aSetValue)
+    {
+        error = Set(SPINEL_PROP_VENDOR_NXP_GET_SET_CCA_CONFIGURE_CMD, SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT8_S
+                    SPINEL_DATATYPE_UINT8_S SPINEL_DATATYPE_UINT8_S SPINEL_DATATYPE_UINT8_S), (&aCcaConfig)->mCcaMode,
+                    (&aCcaConfig)->mCca1Threshold, (&aCcaConfig)->mCca2CorrThreshold, (&aCcaConfig)->mCca2MinNumOfCorrTh);
+    }
+    else
+    {
+        error = Get(SPINEL_PROP_VENDOR_NXP_GET_SET_CCA_CONFIGURE_CMD, SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT8_S
+                    SPINEL_DATATYPE_UINT8_S SPINEL_DATATYPE_UINT8_S SPINEL_DATATYPE_UINT8_S), &aCcaMode, &aCca1Threshold,
+                    &aCca2CorrThreshold, &aCca2MinNumOfCorrTh);
+        (&aCcaConfig)->mCcaMode = aCcaMode;
+        (&aCcaConfig)->mCca1Threshold = aCca1Threshold;
+        (&aCcaConfig)->mCca2CorrThreshold = aCca2CorrThreshold;
+        (&aCcaConfig)->mCca2MinNumOfCorrTh = aCca2MinNumOfCorrTh;
+    }
+
+    return error;
+}
+
+otError RadioSpinel::IRThresholdConfig(otIRConfig &aIRConfig, uint8_t aSetValue)
+{
+   otError error;
+   uint16_t aIRThreshold;
+
+   if(aSetValue)
+   {
+       error = Set(SPINEL_PROP_VENDOR_NXP_IR_CONFIG_THRESHOLD, SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT16_S), (&aIRConfig)->mIRThreshold);
+   }
+   else
+   {
+       error = Get(SPINEL_PROP_VENDOR_NXP_IR_CONFIG_THRESHOLD, SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT16_S), &aIRThreshold);
+       (&aIRConfig)->mIRThreshold = aIRThreshold;
+   }
+   return error;
+}
+
+otError RadioSpinel::GetFwVersion(const char *fwVersion, uint8_t fwVersionLen)
+{
+    otError error = Get(SPINEL_PROP_VENDOR_NXP_GET_FW_VERSION_CMD, SPINEL_DATATYPE_UTF8_S, fwVersion, fwVersionLen);
+    return error;
+}
+
 otError RadioSpinel::SetExtendedAddress(const otExtAddress &aExtAddress)
 {
     otError error;
@@ -1973,7 +2078,39 @@ uint64_t RadioSpinel::GetNow(void) const
    return (mIsTimeSynced) ? (otPlatTimeGet() + mRadioTimeOffset) : UINT64_MAX;
 }
 
-uint32_t RadioSpinel::GetBusSpeed(void) const { return GetSpinelDriver().GetSpinelInterface()->GetBusSpeed(); }
+uint32_t RadioSpinel::GetBusSpeed(void)
+{
+    uint32_t  speed  = 0;
+    Get(SPINEL_PROP_VENDOR_NXP_GET_SET_RCP_FREQUENCY_CMD, SPINEL_DATATYPE_UINT32_S, &speed);
+    return speed;
+}
+
+#ifndef MIN
+#define MIN( x, y ) ( ( x ) < ( y ) ? ( x ) : ( y ) )
+#endif
+
+otError RadioSpinel::SetBusSpeed(uint32_t speed)
+{
+    uint32_t current = GetSpinelDriver().GetSpinelInterface()->GetBusSpeed();
+    otError  error = OT_ERROR_NONE;
+
+    GetSpinelDriver().GetSpinelInterface()->SetBusSpeed(MIN(current, speed));
+
+    error = Set(SPINEL_PROP_VENDOR_NXP_GET_SET_RCP_FREQUENCY_CMD, SPINEL_DATATYPE_UINT32_S, speed);
+
+    if( error == OT_ERROR_NONE )
+    {
+        // Apply new frequency
+        GetSpinelDriver().GetSpinelInterface()->SetBusSpeed(speed);
+    }
+    else
+    {
+        // Back to old frequency
+        GetSpinelDriver().GetSpinelInterface()->SetBusSpeed(current);
+    }
+
+    return error;
+}
 
 uint32_t RadioSpinel::GetBusLatency(void) const { return mBusLatency; }
 
diff --git a/src/lib/spinel/radio_spinel.hpp b/src/lib/spinel/radio_spinel.hpp
index c17503cdb..f55593b90 100644
--- a/src/lib/spinel/radio_spinel.hpp
+++ b/src/lib/spinel/radio_spinel.hpp
@@ -240,6 +240,99 @@ public:
      */
     otError GetIeeeEui64(uint8_t *aIeeeEui64);
 
+    /**
+     * This method sets the factory-assigned IEEE EUI-64 for this transceiver.
+     *
+     * @param[in]  aInstance   The OpenThread instance structure.
+     * @param[out] aIeeeEui64  A pointer to the factory-assigned IEEE EUI-64.
+     *
+     * @retval  OT_ERROR_NONE               Succeeded.
+     * @retval  OT_ERROR_INVALID_STATE      Failed EUI64 is set by factory only except if OTP is not supported.
+     *
+     */
+    otError SetIeeeEui64(const otExtAddress &aIeeeEui64);
+
+    /**
+     * @brief Set the Tx Power Limit for 15.4
+     *
+     * @param txPowerLimit
+     * txPowerLimit (0 or higher than 22), No power backoff is applied
+     * txPowerLimit = 1 to 44, force TX power back off to txPowerLimit
+     * (txPowerLimit = 0.5dBm step, TX power back off : 0.5dBm step )
+     *
+     * @return otError
+     */
+    otError SetTxPowerLimit(uint8_t txPowerLimit);
+
+    /**
+     * @brief Get the Tx Power Limit for 15.4
+     *
+     * @param &txPowerLimit
+     * will contain the value set in IWx12
+     * (txPowerLimit = 0.5dBm step, TX power back off : 0.5dBm step )
+     *
+     * @return otError
+     */
+    otError GetTxPowerLimit(uint8_t &txPowerLimit);
+
+     /**
+     * @brief Set the Independent Reset configuration
+     *
+     * @param &mode
+     * will contain the value set in IWx12
+     * (0, 1, 2)
+     *
+     * @return otError
+     */
+    otError SetIRConfig(uint8_t mode);
+
+     /**
+     * @brief Get the Independent Reset configuration
+     *
+     * @param &mode
+     * will contain the value set in IWx12
+     * (0, 1, 2)
+     *
+     * @return otError
+     */
+    otError GetIRConfig(uint8_t &mode);
+
+    /**
+     * @brief GetMfgChannel
+     * @param channel
+     * @return otError
+     */
+    otError MfgCmd(uint8_t *payload, const uint8_t payloadLenIn, uint8_t &payloadLenOut);
+
+    /**
+     * @brief Set / Get the CCA Mode Configuration Values
+     *
+     * @param &aCcaConfig
+     * will contain the CCA Configuration values of datatype otCCAModeConfig.
+     *
+     * @return otError
+     */
+    otError CcaConfigValue(otCCAModeConfig &aCcaConfig, uint8_t aSetValue);
+
+   /**
+    * This method returns the fw version string.
+    *
+    * @param[out] fwVersion
+    *
+    * @returns otError
+    */
+    otError GetFwVersion(const char *fwVersion, uint8_t fwVersionLen);
+
+    /**
+    * This method sets the threshold time at which OOB toggle will trigger an IR
+    *
+    * @param &aIRConfig
+    * will contain the IR trigger Configuration values of datatype otIRConfig.
+    *
+    * @returns otError
+    */
+    otError IRThresholdConfig(otIRConfig &aIRConfig, uint8_t aSetValue);
+
     /**
      * Sets the Extended Address for address filtering.
      *
@@ -831,7 +924,16 @@ public:
      *
      * @returns   bus speed in bits/second.
      */
-    uint32_t GetBusSpeed(void) const;
+    uint32_t GetBusSpeed(void);
+
+    /**
+     * This method sets the bus speed between the host and the radio.
+     *
+     * @retval  OT_ERROR_NONE           Successfully set the spi frequency.
+     * @retval  OT_ERROR_INVALID_ARGS   frequency is not in valid range.
+     *
+     */
+    otError SetBusSpeed(uint32_t aSpeed);
 
     /**
      * Returns the bus latency between the host and the radio.
diff --git a/src/lib/spinel/spinel.h b/src/lib/spinel/spinel.h
index fb42e30df..2e2f55b41 100644
--- a/src/lib/spinel/spinel.h
+++ b/src/lib/spinel/spinel.h
@@ -5061,6 +5061,17 @@ enum
     SPINEL_PROP_NEST__END = 0x3C00,
 
     SPINEL_PROP_VENDOR__BEGIN = 0x3C00,
+
+    SPINEL_PROP_VENDOR_NXP_IR_CONFIG     = (SPINEL_PROP_VENDOR__BEGIN + 0x100),
+    SPINEL_PROP_VENDOR_NXP_IR_CMD        = (SPINEL_PROP_VENDOR__BEGIN + 0x101),
+    SPINEL_PROP_VENDOR_NXP_SET_EUI64_CMD = (SPINEL_PROP_VENDOR__BEGIN + 0x10A),
+    SPINEL_PROP_VENDOR_NXP_GET_SET_TXPOWERLIMIT_CMD = (SPINEL_PROP_VENDOR__BEGIN + 0x10B),
+    SPINEL_PROP_VENDOR_NXP_GET_SET_CCA_CONFIGURE_CMD = (SPINEL_PROP_VENDOR__BEGIN + 0x10C),
+    SPINEL_PROP_VENDOR_NXP_GET_FW_VERSION_CMD = (SPINEL_PROP_VENDOR__BEGIN + 0x10D),
+    SPINEL_PROP_VENDOR_NXP_GET_SET_RCP_FREQUENCY_CMD = (SPINEL_PROP_VENDOR__BEGIN + 0x10E),
+    SPINEL_PROP_VENDOR_NXP_IR_CONFIG_THRESHOLD = (SPINEL_PROP_VENDOR__BEGIN + 0x10F),
+    SPINEL_CMD_VENDOR_NXP_MFG  = (SPINEL_CMD_VENDOR__BEGIN + 0x3FF),
+
     SPINEL_PROP_VENDOR__END   = 0x4000,
 
     SPINEL_PROP_VENDOR_ESP__BEGIN = (SPINEL_PROP_VENDOR__BEGIN + 0),
diff --git a/src/lib/spinel/spinel_interface.hpp b/src/lib/spinel/spinel_interface.hpp
index ca237b1c6..b8b11acf5 100644
--- a/src/lib/spinel/spinel_interface.hpp
+++ b/src/lib/spinel/spinel_interface.hpp
@@ -120,7 +120,15 @@ public:
      *
      * @returns   Bus speed in bits/second.
      */
-    virtual uint32_t GetBusSpeed(void) const = 0;
+    virtual uint32_t GetBusSpeed(void) = 0;
+
+     /**
+     * This method returns the bus speed between the host and the radio.
+     *
+     * @returns   Bus speed in bits/second.
+     *
+     */
+    virtual void SetBusSpeed(uint32_t aSpeed) = 0;
 
     /**
      * Hardware resets the RCP.
diff --git a/src/posix/main.c b/src/posix/main.c
index 88e373d98..1920d78af 100644
--- a/src/posix/main.c
+++ b/src/posix/main.c
@@ -69,6 +69,16 @@
 #include <openthread/openthread-system.h>
 #include <openthread/platform/misc.h>
 
+extern otError ProcessIRConfig(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessIRCmd(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessSetEui64(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGetSetTxPowerLimit(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessMfgCommands(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGetSetCcaCfg(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGetFwVersion(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGetSetSpiFrequency(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGetSetIRThreshold(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+
 /**
  * Initializes NCP app.
  *
@@ -380,6 +390,15 @@ static const otCliCommand kCommands[] = {
     {"exit", ProcessExit},
 #endif
     {"netif", ProcessNetif},
+    {"ircfg", ProcessIRConfig},    //=> OutOfBand Independent Reset Configuration ircfg <1> means OOB mode
+    {"ircmd", ProcessIRCmd},       //=> InBand Independent Reset command
+    {"seteui64", ProcessSetEui64}, //=> Set ieee.802.15.4 MAC Address
+    {"txpwrlimit", ProcessGetSetTxPowerLimit}, //=> Set TX power limit for 15.4
+    {"mfgcmd", ProcessMfgCommands}, //=> Generic VSC for MFG RF commands
+    {"ccacfg", ProcessGetSetCcaCfg}, //=> Set/Get CCA configuration for 15.4 CCA Before Tx operation
+    {"fwversion", ProcessGetFwVersion}, //=> Get firmware version for 15.4
+    {"spifreq", ProcessGetSetSpiFrequency},  //=> Get/Set SPI frequency supported by RCP
+    {"irthold", ProcessGetSetIRThreshold}   //=> OutOfBand Independent Reset Threshold configuration
 };
 
 int main(int argc, char *argv[])
diff --git a/src/posix/platform/daemon.cpp b/src/posix/platform/daemon.cpp
index b1645e080..1adbf862b 100644
--- a/src/posix/platform/daemon.cpp
+++ b/src/posix/platform/daemon.cpp
@@ -41,6 +41,7 @@
 #include <unistd.h>
 
 #include <openthread/cli.h>
+#include <openthread/thread.h>
 
 #include "cli/cli_config.h"
 #include "common/code_utils.hpp"
@@ -62,6 +63,780 @@ typedef char(Filename)[sizeof(sockaddr_un::sun_path)];
 
 } // namespace
 
+//Vendor Specific commands
+extern "C"
+{
+#define MFG_CMD_ACTION_GET 0
+#define MFG_CMD_ACTION_SET 1
+
+#define MFG_CMD_GET_SET_SPI_FREQUENCY 0x9 // 9
+#define MFG_CMD_GET_SET_CHANNEL 0x0b      // 11
+#define MFG_CMD_GET_SET_TXPOWER 0x0f      // 15
+#define MFG_CMD_CONTINUOUS_TX 0x11        // 17
+#define MFG_CMD_GET_SET_PAYLOAD_SIZE 0x14 // 20
+#define MFG_CMD_GET_RX_RESULT 0x1f        // 31
+#define MFG_CMD_START_RX_TEST 0x20        // 32
+#define MFG_CMD_BURST_TX 0x21             // 33
+#define MFG_CMD_DUTY_CYCLE_TX 0x23        // 35
+#define MFG_CMD_GET_SET_CCA_THRESHOLD  0x2F   // 47
+#define MFG_CMD_CONTINOUS_CCA_TEST 0X31   //49
+#define MFG_CMD_GET_CCA_STATUS 0x32       //50
+#define MFG_CMD_CONTINOUS_ED_TEST 0x37    //55
+#define MFG_CMD_GET_ED_VALUE    0x38      //56
+#define MFG_CMD_PHY_TX_TEST_PSDU 0x39     //57
+#define MFG_CMD_PHY_RX_TX_ACK_TEST 0x3A   //58
+#define MFG_CMD_SET_GENERIC_PARAM 0x3B    //59
+#define MFG_CMD_GENERIC           0xFF    //255
+
+#define MAX_VERSION_STRING_SIZE 128 //< Max size of version string
+#define OT_NXP_PLAT_MFG_FIRST_ANNEX_ID 55
+#define OT_NXP_PLAT_MFG_LAST_ANNEX_ID  108
+
+static uint8_t mfgEnable = 0;
+
+// 15.4_INDEPENDENT_RESET
+otError ProcessIRConfig(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    //syslog(LOG_INFO, "ProcessIRConfig");
+    otError error = OT_ERROR_INVALID_ARGS;
+    uint8_t mode = 0;
+
+    if( aArgsLength == 1 )
+    {
+        mode = (uint8_t)atoi(aArgs[0]);
+        //syslog(LOG_INFO, "-> mode %s", mode==0 ? "Disable IR":mode==3 ? "OOB IR 15.4":(mode==1 ?("OOB IR"):"InBand IR"));
+        if( mode < 4 )
+        {
+            otPlatRadioSetIRConfig((otInstance*)aContext, mode);
+            //syslog(LOG_INFO, "ProcessIRConfig DONE");
+            error = OT_ERROR_NONE;
+        }
+    }
+    else
+    {
+        //syslog(LOG_INFO, "ProcessIRConfig FAILED!");
+        otPlatRadioGetIRConfig((otInstance*)aContext, &mode);
+
+        // Print value as ot-cli output
+        otCliOutputFormat("%d\r\n", mode);
+
+        error = OT_ERROR_NONE;
+    }
+
+    return error;
+}
+
+otError ProcessIRCmd(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgsLength);
+    OT_UNUSED_VARIABLE(aArgs);
+
+    return otPlatRadioSetIRCmd((otInstance*)aContext);
+}
+
+otError ProcessSetEui64(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    otError error = OT_ERROR_INVALID_ARGS;
+
+    if( aArgsLength == 1 )
+    {
+        otExtAddress addr;
+        char        *hex  = *aArgs;
+
+        //syslog(LOG_INFO, "+ SetEui64 %s (len %d)", *aArgs, (uint32_t)strlen(*aArgs));
+
+        if( (hex[1] == 'x') && (strlen(*aArgs) == 18) )
+        {
+            error = OT_ERROR_NONE;
+
+            hex = hex + 2;
+
+            for(uint32_t i = 0; (i < 8) && (error == OT_ERROR_NONE); i++)
+            {
+                addr.m8[i] = 0;
+                for(uint32_t k = 0; k < 2; k++)
+                {
+                    // get current character then increment
+                    uint8_t byte = *hex++;
+                    // transform hex character to the 4bit equivalent number, using the ascii table indexes
+                    if (byte >= '0' && byte <= '9')
+                        byte = byte - '0';
+                    else if (byte >= 'a' && byte <='f')
+                        byte = byte - 'a' + 10;
+                    else if (byte >= 'A' && byte <='F')
+                        byte = byte - 'A' + 10;
+                    else
+                    {
+                        error = OT_ERROR_FAILED;
+                        break;
+                    }
+                    // shift 4 to make space for new digit, and add the 4 bits of the new digit
+                    addr.m8[i] = (addr.m8[i] << 4) | (byte & 0xF);
+                }
+            }
+
+            if( error == OT_ERROR_NONE )
+            {
+                error = otPlatRadioSetIeeeEui64((otInstance*)aContext, (const otExtAddress*)&addr);
+            }
+
+            if( error != OT_ERROR_NONE )
+            {
+                //syslog(LOG_INFO, "- SetEui64 Failed (%#x)", error);
+            }
+            else
+            {
+                //syslog(LOG_INFO, "- SetEui64 SUCCESS");
+            }
+        }
+        else
+        {
+            //syslog(LOG_INFO, "- SetEui64 invalid input arg (0x....?) !");
+        }
+    }
+    else
+    {
+        //syslog(LOG_INFO, "- SetEui64 FAILED !");
+    }
+
+    return error;
+}
+
+otError ProcessGetSetTxPowerLimit(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+
+    otError error = OT_ERROR_INVALID_ARGS;
+    uint8_t txPowerLimit = 0;
+    uint8_t ch26Clamp = 1;
+    //syslog(LOG_INFO, "SetTxPowerLimit");
+
+    if( aArgsLength != 0 ) // set tx power limit
+    {
+        txPowerLimit = (uint8_t)atoi(aArgs[0]);
+
+        if(txPowerLimit < 128)
+        {
+
+            if( aArgsLength == 2 )
+            {
+                if(strcmp(aArgs[1],"disableRFCH26Clamp") == 0)
+                {
+                    ch26Clamp = 0;
+                }
+                else if(strcmp(aArgs[1],"enableRFCH26Clamp") == 0)
+                {
+                    ch26Clamp = 1;
+                }
+
+            }
+
+            txPowerLimit = (txPowerLimit & 0x7F) | (ch26Clamp << 7);
+            otPlatRadioSetTxPowerLimit((otInstance*)aContext, txPowerLimit);
+            //syslog(LOG_INFO, "SetTxPowerLimit DONE");
+            error = OT_ERROR_NONE;
+        }
+        else
+        {
+            error = OT_ERROR_INVALID_ARGS;
+        }
+
+    }
+    else if ( aArgsLength == 0 ) // get tx power limit
+    {
+        otPlatRadioGetTxPowerLimit((otInstance*)aContext, &txPowerLimit);
+
+        // Add value in syslog
+        //syslog(LOG_INFO, "TX power Value value : %d", txPowerLimit);
+
+        // Print value as ot-cli output
+        otCliOutputFormat("Tx Power Limit: %d\r\n", txPowerLimit & 0x7F);
+        otCliOutputFormat("Tx Power Clamping at 0dBm for CH 26: %s\r\n", ((txPowerLimit >> 7) & 0x01)? "enabled":"disabled");
+        error = OT_ERROR_NONE;
+    }
+    else
+    {
+        //syslog(LOG_INFO, "SetTxPowerLimit FAILED! Invalid input arg");
+    }
+
+    return error;
+}
+
+ otError ProcessMfgGetInt8(void *aContext, uint8_t cmdId, uint8_t aArgsLength)
+{
+    otError error = OT_ERROR_INVALID_ARGS;
+    uint8_t outputLen = 0;
+    uint8_t payload[12] = {11};
+    uint8_t payloadLen = 12;
+
+    if(aArgsLength == 1)
+    {
+        payload[1] = cmdId;
+        payload[2] = MFG_CMD_ACTION_GET;
+
+        otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+
+        if((outputLen >= 5) && (payload[3] == 0))
+        {
+            if(cmdId == MFG_CMD_GET_SET_TXPOWER)
+            {
+                otCliOutputFormat("%d\r\n", ((int8_t)payload[4])/2);
+            }
+            else
+            {
+                otCliOutputFormat("%d\r\n", (int8_t)payload[4]);
+            }
+            error = OT_ERROR_NONE;
+        }
+        else{
+            error = OT_ERROR_FAILED;
+        }
+    }
+
+    return error;
+}
+
+ otError ProcessMfgSetInt8(void *aContext, uint8_t cmdId, uint8_t aArgsLength, char *aArgs[], int8_t min, int8_t max)
+{
+    otError error = OT_ERROR_INVALID_ARGS;
+    uint8_t outputLen = 0;
+    uint8_t payload[12] = {11};
+    uint8_t payloadLen = 12;
+    int8_t setValue = 0;
+
+    if(aArgsLength == 2)
+    {
+        setValue = (int8_t)atoi(aArgs[1]);
+        if((setValue >= min) && (setValue <= max))
+        {
+            payload[1] = cmdId;
+            payload[2] = MFG_CMD_ACTION_SET;
+            if(cmdId == MFG_CMD_GET_SET_TXPOWER)
+            {
+                payload[4] = ((uint8_t)setValue) << 1; // convert dBm to half dBm
+            }
+            else
+            {
+                payload[4] = (uint8_t)setValue;
+            }
+
+            otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+
+            if((outputLen >= 4) && (payload[3] == 0))
+            {
+                error = OT_ERROR_NONE;
+            }
+            else
+            {
+                error = OT_ERROR_FAILED;
+            }
+        }
+    }
+
+    return error;
+}
+
+otError ProcessMfgCommands(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    uint8_t payload[12] = {11};
+    uint8_t payloadLen = 12;
+    uint8_t outputLen = 0;
+    otError error = OT_ERROR_INVALID_ARGS;
+    uint8_t cmdId, idx;
+
+    if(aArgsLength == 1)
+    {
+        cmdId = (uint8_t)atoi(aArgs[0]);
+        if((cmdId == 0)||(cmdId == 1))
+        {
+            mfgEnable = cmdId;
+            //syslog(LOG_INFO, "MFG command SUCCESS");
+            return OT_ERROR_NONE;
+        }
+    }
+
+    if(mfgEnable == 0)
+    {
+        //syslog(LOG_INFO, "MFG command not enabled");
+        otCliOutputFormat("MFG command not enabled. to enable it : mfgcmd 1\r\n");
+        return OT_ERROR_NONE;
+    }
+
+    if ((aArgsLength > 0) && (mfgEnable == 1))
+    {
+        cmdId = (uint8_t)atoi(aArgs[0]);
+
+        switch (cmdId)
+        {
+        case MFG_CMD_GET_SET_SPI_FREQUENCY: // get spi freq mode ( 0 => ] 100kHz ;  3.8MHz] ; 1 => ] MIN ; 10MHz] ; not supported )
+                error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_SET_SPI_FREQUENCY, aArgsLength);
+            break;
+
+        case MFG_CMD_GET_SET_SPI_FREQUENCY + 1: // set spi freq ( 0 => ] 100kHz ;  3.8MHz] ; 1 => ] MIN ; 10MHz] )
+                error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_GET_SET_SPI_FREQUENCY, aArgsLength, aArgs, 0, 1);
+            break;
+
+        case MFG_CMD_GET_SET_CHANNEL: // get channel
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_SET_CHANNEL, aArgsLength);
+            break;
+
+        case MFG_CMD_GET_SET_CHANNEL + 1: // set channel
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_GET_SET_CHANNEL, aArgsLength, aArgs, 11, 26);
+            break;
+
+        case MFG_CMD_GET_SET_TXPOWER: // get txpower
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_SET_TXPOWER, aArgsLength);
+            break;
+
+        case MFG_CMD_GET_SET_TXPOWER + 1: // set txpower
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_GET_SET_TXPOWER, aArgsLength, aArgs, -20, 22);
+            break;
+
+        case MFG_CMD_CONTINUOUS_TX:
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_CONTINUOUS_TX, aArgsLength, aArgs, 0, 1);
+            break;
+
+        case MFG_CMD_GET_SET_PAYLOAD_SIZE: // get
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_SET_PAYLOAD_SIZE, aArgsLength);
+            break;
+
+        case MFG_CMD_GET_SET_PAYLOAD_SIZE + 1: // set
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_GET_SET_PAYLOAD_SIZE, aArgsLength, aArgs, 17, 116);
+            break;
+
+        case MFG_CMD_GET_RX_RESULT:
+        {
+            if(aArgsLength == 1)
+            {
+                payload[1] = MFG_CMD_GET_RX_RESULT;
+                payload[2] = MFG_CMD_ACTION_GET;
+                otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                if(outputLen >= 11)
+                {
+                    otCliOutputFormat("status : %d\r\n", payload[4]);
+                    otCliOutputFormat("rx_pkt_count : %d\r\n", payload[5]|(payload[6]<<8));
+                    otCliOutputFormat("total_pkt_count : %d\r\n", payload[7]|(payload[8]<<8));
+                    otCliOutputFormat("rssi : %d\r\n",(int8_t)payload[9]);
+                    otCliOutputFormat("lqi : %d\r\n", payload[10]);
+                    error = OT_ERROR_NONE;
+                }
+                else{
+                    error = OT_ERROR_FAILED;
+                }
+            }
+        }
+        break;
+
+        case MFG_CMD_START_RX_TEST:
+        {
+            if(aArgsLength == 1)
+            {
+                payload[1] = MFG_CMD_START_RX_TEST;
+                otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                error = OT_ERROR_NONE;
+            }
+        }
+        break;
+
+        case MFG_CMD_BURST_TX:
+        {
+            uint8_t mode = 0, gap = 0;
+            if(aArgsLength == 3)
+            {
+                mode = (uint8_t)atoi(aArgs[1]);
+                gap = (uint8_t)atoi(aArgs[2]);
+                if((mode < 8) && (gap > 5))
+                {
+                    payload[1] = MFG_CMD_BURST_TX;
+                    payload[4] = mode;
+                    payload[5] = gap;
+                    otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                    error = OT_ERROR_NONE;
+                }
+            }
+        }
+        break;
+
+        case MFG_CMD_DUTY_CYCLE_TX:
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_DUTY_CYCLE_TX, aArgsLength, aArgs, 0, 1);
+            break;
+
+        case MFG_CMD_GET_SET_CCA_THRESHOLD: // get
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_SET_CCA_THRESHOLD, aArgsLength);
+            break;
+
+        case MFG_CMD_GET_SET_CCA_THRESHOLD + 1: // set
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_GET_SET_CCA_THRESHOLD, aArgsLength, aArgs, -110, 0);
+            break;
+
+        case MFG_CMD_CONTINOUS_CCA_TEST:
+        {
+            if(aArgsLength == 3)
+            {
+                payload[1] = MFG_CMD_CONTINOUS_CCA_TEST;
+                payload[2] = MFG_CMD_ACTION_SET;
+                payload[4] = (uint8_t)atoi(aArgs[1]); // enable
+                payload[5] = (uint8_t)atoi(aArgs[2]); // CCA Mode
+                if((payload[4] < 2) && (payload[5] < 4))
+                {
+                    otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                    if((outputLen >= 4) && (payload[3] == 0))
+                    {
+                        error = OT_ERROR_NONE;
+                    }
+                    else{
+                        error = OT_ERROR_FAILED;
+                    }
+                }
+            }
+        }
+        break;
+
+        case MFG_CMD_GET_CCA_STATUS: // get
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_CCA_STATUS, aArgsLength);
+            break;
+
+        case MFG_CMD_CONTINOUS_ED_TEST:
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_CONTINOUS_ED_TEST, aArgsLength, aArgs, 0, 1);
+            break;
+
+        case MFG_CMD_GET_ED_VALUE:
+            error = ProcessMfgGetInt8((otInstance*)aContext, MFG_CMD_GET_ED_VALUE, aArgsLength);
+            break;
+
+        case MFG_CMD_PHY_TX_TEST_PSDU:
+        {
+            uint8_t count_opt, gap, ackEnable;
+            if(aArgsLength == 4)
+            {
+                payload[1]  = MFG_CMD_PHY_TX_TEST_PSDU;
+                payload[2]  = MFG_CMD_ACTION_SET;
+
+                count_opt = (uint8_t)atoi(aArgs[1]);
+                gap       = (uint8_t)atoi(aArgs[2]);
+                ackEnable = (uint8_t)atoi(aArgs[3]);
+                if((count_opt < 8) && (gap > 5) && (ackEnable < 2))
+                {
+                    payload[4]  = count_opt;
+                    payload[5]  = gap;
+                    payload[6]  = ackEnable;
+                    otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                    error = OT_ERROR_NONE;
+                }
+            }
+        }
+        break;
+
+        case MFG_CMD_PHY_RX_TX_ACK_TEST:
+            error = ProcessMfgSetInt8((otInstance*)aContext, MFG_CMD_PHY_RX_TX_ACK_TEST, aArgsLength, aArgs, 0, 1);
+            break;
+
+        case MFG_CMD_SET_GENERIC_PARAM:
+        {
+            uint16_t panid, destaddr, srcaddr;
+            if(aArgsLength == 5)
+            {
+                panid       = (uint16_t)strtol(aArgs[2], NULL, 16);
+                destaddr    = (uint16_t)strtol(aArgs[3], NULL, 16);
+                srcaddr     = (uint16_t)strtol(aArgs[4], NULL, 16);
+
+                payload[1]  = MFG_CMD_SET_GENERIC_PARAM;
+                payload[2]  = MFG_CMD_ACTION_SET;
+                payload[4]  = (uint8_t) atoi(aArgs[1]); // SEQ_NUM
+                payload[5]  = (uint8_t) (panid & 0xFF); // PAN ID LSB
+                payload[6]  = (uint8_t) ((panid >> 8) & 0xFF); // PAN ID MSB
+                payload[7]  = (uint8_t) (destaddr & 0xFF); // DEST ADDR LSB
+                payload[8]  = (uint8_t) ((destaddr >> 8) & 0xFF); // DEST ADDR MSB
+                payload[9]  = (uint8_t) (srcaddr & 0xFF); // SRC ADDR LSB
+                payload[10] = (uint8_t) ((srcaddr >> 8) & 0xFF); // SRC ADDR MSB
+
+                otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload, payloadLen, &outputLen);
+                error = OT_ERROR_NONE;
+            }
+        }
+        break;
+
+        case MFG_CMD_GENERIC:
+        {
+            uint8_t Action       = (uint8_t) atoi(aArgs[1]);
+            uint8_t AnnexVal     = (uint8_t) atoi(aArgs[2]);
+            char *Annexdata      = aArgs[3];
+            uint8_t DataLen      = strlen(Annexdata);
+            uint8_t PayloadLen   = 0;
+            uint8_t payload_generic_cmd [256];
+
+            if (Action == MFG_CMD_ACTION_SET)
+            {
+                if ( (AnnexVal < OT_NXP_PLAT_MFG_FIRST_ANNEX_ID) && (AnnexVal > OT_NXP_PLAT_MFG_LAST_ANNEX_ID))
+                {
+                    otLogInfoPlat("The Annex Type limit set is out of range");
+                }
+                else if (DataLen %2 != 0)
+                {
+                    otLogInfoPlat("The Annex Data input format allows only in Bytes format.");
+                }
+                else if ((DataLen > 2) && (Annexdata[0] == '0') && (Annexdata[1] == 'x'))
+                {
+                    DataLen   -= 2;
+                    Annexdata += 2;
+
+                    DataLen    = DataLen / 2; //Data length in bytes.
+                    PayloadLen = DataLen + 6;
+
+                    payload_generic_cmd [0] = PayloadLen - 1 ;            // size of the payload
+                    payload_generic_cmd [1] = MFG_CMD_GENERIC;            // command id
+                    payload_generic_cmd [2] = Action;                     // action set or get
+                    payload_generic_cmd [3] = 0;                          // return value
+                    payload_generic_cmd [4] = AnnexVal;                   // sub commands. for Annex100 it is 100.
+                    payload_generic_cmd [5] = DataLen;                    // Annex data len
+
+                    for ( uint8_t index = 0; index < DataLen; index++)
+                    {
+                        char byte_str[3] = {Annexdata[ 2* index], Annexdata[ 2* index + 1], '\0'};
+                        payload_generic_cmd[6 + index] = (unsigned int)strtoul(byte_str, NULL,16);
+                    }
+
+                    otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t*)payload_generic_cmd, PayloadLen , &outputLen);
+
+                    for(idx = 0; idx < outputLen; idx++)
+                    {
+                        otCliOutputFormat("%d ", payload_generic_cmd[idx]);
+                    }
+                    otCliOutputFormat("\r\n");
+
+                    if (payload_generic_cmd[3] == 0)
+                    {
+                        otLogInfoPlat("Setting Annex%d success", AnnexVal);
+                        error = OT_ERROR_NONE;
+                    }
+                    else if (payload_generic_cmd[3] == 0xff)
+                    {
+                        error = OT_ERROR_INVALID_ARGS;
+                    }
+                    else
+                    {
+                        error = OT_ERROR_FAILED;
+                    }
+                }
+                else
+                {
+                    otLogInfoPlat("The Annex Data input format allows only Hexadecimal.");
+                }
+            }
+            else if (Action == MFG_CMD_ACTION_GET)
+            {
+                error = OT_ERROR_NOT_IMPLEMENTED;
+            }
+        }
+        break;
+
+	default:
+            error = OT_ERROR_NOT_IMPLEMENTED;
+            break;
+        }
+    }
+
+    //HANDLE ERRORS
+    if(error == OT_ERROR_NONE)
+    {
+        //syslog(LOG_INFO, "MFG command SUCCESS");
+    }
+    else if(aArgsLength == payloadLen)
+    {
+        // If user passed all the payload, this means this is a direct message for the RCP.
+        // Send it and print the return results.
+        for(idx = 0; idx < payloadLen; idx++)
+        {
+            payload[idx] = (uint8_t)atoi(aArgs[idx]);
+        }
+        otPlatRadioMfgCommand((otInstance*)aContext, (uint8_t *)payload, payloadLen, &outputLen);
+        for(idx = 0; idx < outputLen; idx++)
+        {
+            otCliOutputFormat("%d ", payload[idx]);
+        }
+        otCliOutputFormat("\r\n");
+        error = OT_ERROR_NONE;
+        //syslog(LOG_INFO, "MFG command SUCCESS");
+    }
+    else if(error == OT_ERROR_INVALID_ARGS)
+    {
+        //syslog(LOG_INFO, "MFG command Invalid parameter");
+        //otCliOutputFormat("INVALID PARAMETER\r\n");
+    }
+    else if(error == OT_ERROR_NOT_IMPLEMENTED)
+    {
+        //syslog(LOG_INFO, "MFG command not implemented");
+        otCliOutputFormat("NOT IMPLEMENTED\r\n");
+    }
+    else
+    {
+        //syslog(LOG_INFO, "MFG command FAILED");
+        otCliOutputFormat("FAILED\r\n");
+    }
+
+    return error;
+}
+
+otError ProcessGetSetCcaCfg(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    otCCAModeConfig aCcaCfg;
+    //syslog(LOG_INFO, "SetCcaConfig");
+
+    if( aArgsLength == 4 ) // set cca configuration
+    {
+        aCcaCfg.mCcaMode            = (uint8_t)strtol(aArgs[0], NULL, 16);
+        aCcaCfg.mCca1Threshold      = (uint8_t)strtol(aArgs[1], NULL, 16);
+        aCcaCfg.mCca2CorrThreshold  = (uint8_t)strtol(aArgs[2], NULL, 16);
+        aCcaCfg.mCca2MinNumOfCorrTh = (uint8_t)strtol(aArgs[3], NULL, 16);
+        if ((((aCcaCfg.mCcaMode >= 1) && (aCcaCfg.mCcaMode <= 4)) || (aCcaCfg.mCcaMode == 0xFF)) &&
+           (aCcaCfg.mCca2MinNumOfCorrTh <= 6) ){
+            otPlatRadioCcaConfigValue((otInstance*)aContext, &aCcaCfg, 0x1);
+            //syslog(LOG_INFO, "SetCcaConfig DONE");
+        }
+    }
+    else if ( aArgsLength == 0 ) // get cca configuration
+    {
+        otPlatRadioCcaConfigValue((otInstance*)aContext, &aCcaCfg, 0x0);
+
+        // Add value in syslog
+        //syslog(LOG_INFO, "CCA Configuration:\r\n");
+        //syslog(LOG_INFO, "CCA Mode type: CCA1=0x01, CCA2=0x02, CCA3=0x03[CCA1 AND CCA2], CCA3=0x04[CCA1 OR CCA2], NoCCA=0xFF: 0x%x\r\n", aCcaCfg.mCcaMode);
+        //syslog(LOG_INFO, "CCA1 Threshold Value : 0x%x\r\n", aCcaCfg.mCca1Threshold);
+        //syslog(LOG_INFO, "CCA2 Correlation Threshold Value : 0x%x\r\n", aCcaCfg.mCca2CorrThreshold);
+        //syslog(LOG_INFO, "CCA2 Minimim Number of Correlation Threshold Value : 0x%x\r\n", aCcaCfg.mCca2MinNumOfCorrTh);
+
+        // Print value as ot-cli output
+        otCliOutputFormat("CCA Configuration:\r\n");
+        otCliOutputFormat("CCA Mode type: CCA1=0x01, CCA2=0x02, CCA3=0x03[CCA1 AND CCA2], CCA3=0x04[CCA1 OR CCA2], NoCCA=0xFF: 0x%x\r\n", aCcaCfg.mCcaMode);
+        otCliOutputFormat("CCA1 Threshold Value : 0x%x\r\n", aCcaCfg.mCca1Threshold);
+        otCliOutputFormat("CCA2 Correlation Threshold Value : 0x%x\r\n", aCcaCfg.mCca2CorrThreshold);
+        otCliOutputFormat("CCA2 Minimim Number of Correlation Threshold Value : 0x%x\r\n", aCcaCfg.mCca2MinNumOfCorrTh);
+    }
+    else
+    {
+        //syslog(LOG_INFO, "ccacfg FAILED! Invalid input arg\r\nFormat: ccacfg <CcaMode> <Cca1Threshold> <Cca2CorrThreshold> <Cca2MinNumOfCorrTh>\r\nCcaMode: CCA Mode type [CCA1=0x01, CCA2=0x02, CCA3=0x03[CCA1 AND CCA2], CCA3=0x04[CCA1 OR CCA2], NoCCA=0xFF]\r\nCca1Threshold[1Byte Hex value]: Energy threshold for CCA Mode1\r\nCca2CorrThreshold[1Byte Hex value]: CCA Mode 2 Correlation Threshold\r\nCca2MinNumOfCorrTh: [0 to 6]\r\n");
+        otCliOutputFormat("ccacfg FAILED! Invalid input arg\r\nFormat: ccacfg <CcaMode> <Cca1Threshold> <Cca2CorrThreshold> <Cca2MinNumOfCorrTh>\r\nCcaMode: CCA Mode type [CCA1=0x01, CCA2=0x02, CCA3=0x03[CCA1 AND CCA2], CCA3=0x04[CCA1 OR CCA2], NoCCA=0xFF]\r\nCca1Threshold[1Byte Hex value]: Energy threshold for CCA Mode1\r\nCca2CorrThreshold[1Byte Hex value]: CCA Mode 2 Correlation Threshold\r\nCca2MinNumOfCorrTh: [0 to 6]\r\n");
+    }
+
+    return OT_ERROR_NONE;
+}
+
+otError ProcessGetFwVersion(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgs);
+
+    if ( aArgsLength == 0 )
+    {
+        const char version[MAX_VERSION_STRING_SIZE] = {0};
+        otPlatRadioGetFwVersionString((otInstance*)aContext, version, MAX_VERSION_STRING_SIZE);
+        otCliOutputFormat("%s\r\n", version);
+    }
+    else
+    {
+        //syslog(LOG_INFO, "GetFwVersion FAILED! Invalid input arg");
+    }
+    return OT_ERROR_NONE;
+}
+
+otError ProcessGetSetSpiFrequency(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgs);
+
+    if ( aArgsLength == 0 )
+    {
+        uint32_t speed = otPlatRadioGetBusSpeed((otInstance*)aContext);
+
+        switch(speed)
+        {
+            case 1: //OT_RADIO_HIGH_BUS_SPEED
+                otCliOutputFormat("Get => RCP supports SPI Bus Speed until 10MHz\r\n");
+            break;
+            case 2: //OT_RADIO_FULL_BUS_SPEED
+                otCliOutputFormat("Get => RCP supports SPI Bus Speed > 10MHz\r\n");
+            break;
+            case 0: //OT_RADIO_DEFAULT_BUS_SPEED
+            default:
+                otCliOutputFormat("Get => RCP supports SPI Bus Speed until 4MHz\r\n");
+            break;
+        }
+    }
+    else if ( aArgsLength == 1 )
+    {
+        uint32_t speed = (uint32_t)atoi(aArgs[0]);
+
+        if( (speed >= 10000) && (speed <= 10000000) )
+        {
+            if( otThreadGetDeviceRole((otInstance*)aContext) == OT_DEVICE_ROLE_DISABLED )
+            {
+                if( otPlatRadioSetBusSpeed((otInstance*)aContext, speed) != OT_ERROR_NONE )
+                {
+                    otCliOutputFormat("Get/Set SpiFrequency FAILED! Invalid input <speed %d>\n", speed);
+                }
+            }
+            else
+            {
+                otCliOutputFormat("Get/Set SpiFrequency FAILED! Invalid STATE - %s -\n", otThreadDeviceRoleToString(otThreadGetDeviceRole((otInstance*)aContext)));
+            }
+        }
+        else
+        {
+            otCliOutputFormat("Set SpiFrequency FAILED! Invalid speed range - required [10000 (10kHz) : 10000000 (10MHz)]>\n");
+        }
+    }
+    else
+    {
+        otCliOutputFormat("Get/Set SpiFrequency FAILED! Invalid input <speed>\n");
+    }
+    return OT_ERROR_NONE;
+}
+
+otError ProcessGetSetIRThreshold(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    otIRConfig aIRConfig;
+    //syslog(LOG_INFO, "ProcessGetSetIRThreshold");
+
+    if( aArgsLength == 1 )      // Set IR threshold
+    {
+        aIRConfig.mIRThreshold = (uint16_t)atoi(aArgs[0]);
+        if( ( aIRConfig.mIRThreshold >= 100 ) && ( aIRConfig.mIRThreshold<= 1000 ) )
+        {
+           otPlatRadioIRThresholdConfig((otInstance*)aContext, &aIRConfig, 0x01);
+           //syslog(LOG_INFO, "OOB IR Threshold: %d\r\n", aIRConfig.mIRThreshold);
+        }
+        else
+        {
+           //syslog(LOG_INFO, "OOB IR Threshold FAILED! Invalid Threshold Time - Required[100 to 1000\r\n");
+           otCliOutputFormat("OOB IR Threshold FAILED! Invalid Threshold Time - Required[100 to 1000\r\n");
+        }
+    }
+
+    else if ( aArgsLength == 0 ) // get IR threshold
+    {
+        otPlatRadioIRThresholdConfig((otInstance*)aContext, &aIRConfig, 0x00);
+        //syslog(LOG_INFO, "OOB IR Threshold: %d\r\n", aIRConfig.mIRThreshold);
+        otCliOutputFormat("OOB IR Threshold: %d\r\n", aIRConfig.mIRThreshold);
+    }
+
+    else
+    {
+        //syslog(LOG_INFO, "OOB IR Threshold FAILED! Invalid input arg\r\nFormat: irthold <Threshold Time>\r\nThreshold Time : 100 to 1000\r\n");
+        otCliOutputFormat("OOB IR Threshold FAILED! Invalid input arg\r\nFormat: irthold <Threshold Time>\r\nThreshold Time : 100 to 1000\r\n");
+    }
+    return OT_ERROR_NONE;
+}
+
+static const otCliCommand kCommands[] = {
+    {"ircfg", ProcessIRConfig},    //=> OutOfBand Independent Reset Configuration ircfg <1> means OOB mode
+    {"ircmd", ProcessIRCmd},       //=> InBand Independent Reset command
+    {"seteui64", ProcessSetEui64}, //=> Set ieee.802.15.4 MAC Address
+    {"txpwrlimit", ProcessGetSetTxPowerLimit}, //=> Set TX power limit for 15.4
+    {"mfgcmd", ProcessMfgCommands}, //=> Generic VSC for MFG RF commands
+    {"ccacfg", ProcessGetSetCcaCfg}, //=> Set/Get CCA configuration for 15.4 CCA Before Tx operation
+    {"fwversion", ProcessGetFwVersion}, //=> Get firmware version for 15.4
+    {"spifreq", ProcessGetSetSpiFrequency},  //=> Get/Set SPI frequency supported by RCP
+    {"irthold", ProcessGetSetIRThreshold}   //=> OutOfBand Independent Reset Threshold configuration
+};
+} //extern "C"
+
 // using macro to avoid the warning about format-nonliteral
 #define GetFilename(aFilename, aPattern)                                                                       \
     do                                                                                                         \
@@ -288,6 +1063,8 @@ exit:
     otSysCliInitUsingDaemon(gInstance);
 #endif
 
+    otCliSetUserCommands(kCommands, OT_ARRAY_LENGTH(kCommands), gInstance);
+
     Mainloop::Manager::Get().Add(*this);
 
     return;
diff --git a/src/posix/platform/hdlc_interface.hpp b/src/posix/platform/hdlc_interface.hpp
index 9c376162c..5d8356a5f 100644
--- a/src/posix/platform/hdlc_interface.hpp
+++ b/src/posix/platform/hdlc_interface.hpp
@@ -129,7 +129,13 @@ public:
      *
      * @returns   Bus speed in bits/second.
      */
-    uint32_t GetBusSpeed(void) const { return mBaudRate; }
+    uint32_t GetBusSpeed(void) { return mBaudRate; }
+
+    /**
+     * This method sets the bus speed between the host and the radio.
+     *
+     */
+    void SetBusSpeed(uint32_t aSpeed) { OT_UNUSED_VARIABLE(aSpeed); }
 
     /**
      * Hardware resets the RCP.
diff --git a/src/posix/platform/radio.cpp b/src/posix/platform/radio.cpp
index 4a525e7d1..2a539c52a 100644
--- a/src/posix/platform/radio.cpp
+++ b/src/posix/platform/radio.cpp
@@ -981,6 +981,12 @@ uint32_t otPlatRadioGetBusSpeed(otInstance *aInstance)
     return GetRadioSpinel().GetBusSpeed();
 }
 
+otError otPlatRadioSetBusSpeed(otInstance *aInstance, uint32_t aSpeed)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().SetBusSpeed(aSpeed);
+}
+
 uint32_t otPlatRadioGetBusLatency(otInstance *aInstance)
 {
     OT_UNUSED_VARIABLE(aInstance);
@@ -1097,6 +1103,75 @@ otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel, uint32_t a
     return OT_ERROR_NOT_IMPLEMENTED;
 }
 
+otError otPlatRadioSetIRConfig(otInstance *aInstance, uint8_t mode)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().SetIRConfig(mode);
+}
+
+otError otPlatRadioGetIRConfig(otInstance *aInstance, uint8_t* mode)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().GetIRConfig(*mode);
+}
+
+otError otPlatRadioSetIRCmd(otInstance *aInstance)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    otError status = GetRadioSpinel().Set(SPINEL_PROP_VENDOR_NXP_IR_CMD, SPINEL_DATATYPE_UINT16_S);
+    if (status == OT_ERROR_NONE) DieNow(status);
+    return status;
+}
+
+otError otPlatRadioSetIeeeEui64(otInstance *aInstance, const otExtAddress *aIeeeEui64)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    otExtAddress addr;
+
+    for (size_t i = 0; i < sizeof(addr); i++)
+    {
+        addr.m8[i] = aIeeeEui64->m8[sizeof(addr) - 1 - i];
+    }
+
+    return GetRadioSpinel().SetIeeeEui64(*aIeeeEui64);
+}
+
+otError otPlatRadioSetTxPowerLimit(otInstance *aInstance, uint8_t txPowerLimit)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().SetTxPowerLimit(txPowerLimit);
+}
+
+otError otPlatRadioGetTxPowerLimit(otInstance *aInstance, uint8_t* txPowerLimit)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().GetTxPowerLimit(*txPowerLimit);
+}
+
+otError otPlatRadioMfgCommand(otInstance *aInstance, uint8_t *payload, const uint8_t payloadLenIn, uint8_t *payloadLenOut)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().MfgCmd(payload, payloadLenIn, *payloadLenOut);
+}
+
+otError otPlatRadioCcaConfigValue(otInstance *aInstance, otCCAModeConfig *aOtCCAModeConfig, uint8_t aSetValue)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().CcaConfigValue(*aOtCCAModeConfig, aSetValue);
+}
+
+otError otPlatRadioGetFwVersionString(otInstance *aInstance, const char *fwVersion, uint8_t fwVersionLen)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().GetFwVersion(fwVersion, fwVersionLen);
+}
+
+otError otPlatRadioIRThresholdConfig(otInstance *aInstance, otIRConfig *aotIRConfig, uint8_t aSetValue)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    return GetRadioSpinel().IRThresholdConfig(*aotIRConfig, aSetValue);
+}
+
 #if OPENTHREAD_CONFIG_PLATFORM_BOOTLOADER_MODE_ENABLE
 otError otPlatResetToBootloader(otInstance *aInstance)
 {
diff --git a/src/posix/platform/spi_interface.hpp b/src/posix/platform/spi_interface.hpp
index 5524b5b01..0d81bc86e 100644
--- a/src/posix/platform/spi_interface.hpp
+++ b/src/posix/platform/spi_interface.hpp
@@ -130,7 +130,15 @@ public:
      *
      * @returns   Bus speed in bits/second.
      */
-    uint32_t GetBusSpeed(void) const { return ((mSpiDevFd >= 0) ? mSpiSpeedHz : 0); }
+    uint32_t GetBusSpeed(void) { return ((mSpiDevFd >= 0) ? mSpiSpeedHz : 0); }
+
+    /**
+     * This method returns the bus speed between the host and the radio.
+     *
+     * @returns   Bus speed in bits/second.
+     *
+     */
+    void SetBusSpeed(uint32_t aSpeed) { ((mSpiDevFd >= 0) ? mSpiSpeedHz = aSpeed : mSpiSpeedHz = 0); }
 
     /**
      * Hardware resets the RCP.
